lib/internal/async_hooks.js

```diff
@@ -332,6 +339,39 @@ function emitDestroyScript(asyncId) {
 }
 
 
+// This is the equivalent of the native push_async_ids() call.
+function pushAsyncIds(asyncId, triggerAsyncId) {
+  const stackLength = async_hook_fields[kStackLength];
+  if (stackLength >= async_hook_fields[kStackCapacity])
+    return pushAsyncIds_(asyncId, triggerAsyncId);
+  const offset = stackLength;
+  async_wrap.async_ids_stack[offset * 2] = async_id_fields[kExecutionAsyncId];
+  async_wrap.async_ids_stack[offset * 2 + 1] = async_id_fields[kTriggerAsyncId];
+  async_hook_fields[kStackLength]++;
+  async_id_fields[kExecutionAsyncId] = asyncId;
+  async_id_fields[kTriggerAsyncId] = triggerAsyncId;
+}
```

Comment:
I'd take a slightly different approach here.  Growing the stack can be done in JS: allocate a new typed array, copy over the elements, then call into C++ to update the AliasedBuffer's reference.  That probably also obviates the need to overload `operator=(AliasedBuffer&&)`.

If you wrote it this way because of how PromiseHook() calls `env->async_hooks()->push_async_ids()`, perhaps you can work around that by ensuring there are always a few free slots?  I think that should work although admittedly I didn't look too closely.

Reply:
@bnoordhuis It’s not just `PromiseHook()`, `MakeCallback()` also does that – I think for your suggestion to work in a generic fashion you’d have to call that growing function from C++, which kind of just moves the problem around